### PR TITLE
Update pinned nightly to fix `cargo udeps` checks

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context
CI started failing due to the release of `lock_api-0.4.7` (see [failed build here](https://github.com/awslabs/smithy-rs/runs/5759821814)), which does not appear to compile against `nightly-2022-03-03`. This PR upgrades the nightly to a version that it successfully compiles against.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/aws-smithy-types/additional-ci [hunk 1]
```diff
@@ -12,4 +12,4 @@ echo "### Checking for duplicate dependency versions in the normal dependency gr
 cargo tree -d --edges normal --all-features
 
 echo "### Running api-linter"
-cargo +nightly-2022-03-03 api-linter --all-features
+cargo "+${RUST_NIGHTLY_VERSION}" api-linter --all-features
```
Comment (rcoh) [inline rust-runtime/aws-smithy-types/additional-ci:15]: Where does this var come from?

**Review by rcoh — APPROVED** [review]

File: rust-runtime/aws-smithy-types/additional-ci (hunk 1 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-types/additional-ci:15]: It's an environment variable set in the build Docker image here: https://github.com/awslabs/smithy-rs/blob/main/tools/ci-build/base-image.dockerfile#L117

It just occurred to me that using it here will mean we can't directly invoke these additional-ci scripts anymore... What's your preference on having more locations to update vs. having to jump through hoops to run one of these test scripts? I'm leaning towards having more places to update until it's easier to run the Docker CI actions.

Comment (rcoh) [reply rust-runtime/aws-smithy-types/additional-ci:15]: I assumed we'd be using docker to run the actions locally. You can also define a default in this file with something like:
`RUST_NIGHTLY_VERSION=${blah:-$RUST_NIGHTLY_VERSION}`

**Review by rcoh — APPROVED** [review]

File: tools/Dockerfile [hunk 2]
```diff
@@ -36,10 +39,12 @@ RUN set -eux; \
...
 FROM bare_base_image AS install_rust
 ARG rust_stable_version=1.56.1
-ARG rust_nightly_version=nightly-2022-03-03
+ARG rust_nightly_version=nightly-2022-03-29
 ARG cargo_udeps_version=0.1.27
 ARG cargo_hack_version=0.5.12
-ARG smithy_rs_revision=main
+# If the `checkout_smithy_rs_tools` arg is set to true, then the Dockerfile will acquire the tools
+# source code by checking out awslabs/smithy-rs/main rather than copying them from the local directory.
+ARG checkout_smithy_rs_tools=false
```
Comment (rcoh) [inline tools/Dockerfile:47]: do we even need this arg 🤔 

We could have the other docker build handle the repo clone separately from building...

Comment (rcoh) [inline tools/Dockerfile:47]: do we even need this arg 🤔 

We could have the other docker build handle the repo clone separately from building...

**Review by rcoh — APPROVED** [review]

**Pull request merged**